packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+    if (progress < 0.5) {
+      return begin;
+    } else {
+      return end;
+    }
+  }
+
+  @override
+  Animation<double> getScaleAnimation({Animation<double> parent}) {
+    // Animate the scale down from 1 to 0 in the first half of the animation
+    // then from 0 back to 1 in the second half.
+    final Curve curve = const Interval(0.5, 1.0, curve: Curves.ease);
+    return new AnimationMax<double>(
+      new ReverseAnimation(new CurveTween(curve: curve.flipped).animate(parent)),
+      new CurveTween(curve: curve).animate(parent),
+    );
+  }
+
+  @override
+  Animation<double> getRotationAnimation({Animation<double> parent}) {
+    // Because we only see the last half of the rotation tween, 
+    // it needs to go twice as far.
+    final Tween<double> rotationTween = new Tween<double>(
+      begin: 1.0 - _kFloatingActionButtonTurnInterval * 2, 
+      end: 1.0,
+    );
+    // This rotation will turn on the way in, but not on the way out.
+    return new AnimationMax<double>(
+      rotationTween.animate(parent),
+      new ReverseAnimation(new CurveTween(curve: const Threshold(0.5)).animate(parent)),
+    );
+  }
+
+  // If the animation was just starting, we'll continue from where we left off.
+  // If the animation was finishing, we'll treat it as if we were starting at that point in reverse.
+  // This avoids a size jump during the animation.
+  @override
+  double getAnimationRestart(double previousValue) => math.min(1.0 - previousValue, previousValue);
+}
+
+/// The geometry of the [Scaffold] before it finishes laying out.
+/// 
+/// The Scaffold passes this prelayout geometry to its [FloatingActionButtonPositioner].
+/// 
+/// For a description of the [Scaffold]'s geometry after it has
+/// finished laying out, see the [ScaffoldGeometry].
+@immutable
+class ScaffoldPrelayoutGeometry {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const ScaffoldPrelayoutGeometry({this.bottomSheetSize, this.contentBottom, this.contentTop, this.floatingActionButtonSize, this.horizontalFloatingActionButtonPadding, this.scaffoldSize, this.snackBarSize, this.textDirection});
+
+  /// The [Size] of the [Scaffold]'s [FloatingActionButton] (if available).
+  /// 
+  /// The [Scaffold] will determine the [floatingActionButtonSize] if 
+  /// [Scaffold.floatingActionButton] is not null.
+  /// 
+  /// If [Scaffold.floatingActionButton] is null, this should be [Size.zero].
+  final Size floatingActionButtonSize;
+
+  /// The [Size] of the [Scaffold]'s [BottomSheet] (if available).
+  /// 
+  /// The [Scaffold] will determine the [bottomSheetSize] if the [Scaffold]
+  /// is currently showing a [BottomSheet].
+  /// 
+  /// If the [Scaffold] is not currently showing a [BottomSheet],
+  /// this should be [Size.zero].
+  final Size bottomSheetSize;
+
+  /// The height from the [Scaffold]'s top where its body ends.
```

Comment:
I think this means: the vertical distance from the Scaffold's origin to the bottom of its [Scaffold.body]. 

It would be helpful to add a sentence or two about why this value is useful and why it's not bottomSheetSize.height.